lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -843,25 +841,48 @@ assembleTransaction
                         _ -> throwError $ InvalidUtxoInputAmount input
                 pure (TxIn txHash txIndex, coin)
 
-unmarshalMetadataValue :: Aeson.Value -> Either String TxMetadataValue
-unmarshalMetadataValue = \case
-    Aeson.Object km ->
-        TxMetaMap <$> for
-            (Aeson.toList km)
-            (bitraverse (unmarshalMetadataValue . Aeson.String . Aeson.toText)
-                unmarshalMetadataValue)
-    Aeson.Array vec ->
-        TxMetaList . V.toList <$> for vec unmarshalMetadataValue
-    Aeson.String txt ->
-        Right $ TxMetaText txt
-    Aeson.Number sci ->
-        if isInteger sci
-            then Right (TxMetaNumber (truncate sci))
-            else Left "Non-integer metadata value"
-    Aeson.Bool b ->
-        Left $ "Expected TxMetadataValue but got bool (" <> show b <> ")"
-    Aeson.Null ->
-        Left "Expected TxMetadataValue but got null"
+unmarshalMetadata ::
+    [BF.TransactionMetaCBOR] -> Either BlockfrostError TxMetadata
+unmarshalMetadata cbor = TxMetadata . Map.fromList <$> do
+    mbValues <- for cbor \BF.TransactionMetaCBOR{..} ->
+        for _transactionMetaCBORMetadata
+            (first MetadataError . unmarshalMetadataValue)
+    for (catMaybes mbValues) $ \case
+       TxMetaMap [(TxMetaNumber n, v)] | n >= 0 -> pure (fromIntegral n, v)
+       value -> throwError . MetadataError $ InvalidTxMetadataValue value
+
+unmarshalMetadataValue :: Text -> Either BlockfrostMetadataError TxMetadataValue
+unmarshalMetadataValue txt = do
+    hex <- first HexDecodingError . convertFromBase Base16 $ encodeUtf8 txt
+    term <- first DeserialisationError . deserialiseOrFail $ fromStrict hex
+    unmarshalMetadataTerm term
+
+unmarshalMetadataTerm ::
+    CBOR.Term -> Either BlockfrostMetadataError TxMetadataValue
+unmarshalMetadataTerm = \case
+    TInt n -> Right $ TxMetaNumber $ fromIntegral n
+    TInteger n -> Right $ TxMetaNumber n
+    TBytes bs -> Right $ TxMetaBytes $ convertToBase Base16 bs
```

Comment:
Actually, are you sure about this `Base16` thing? 🤔

Reply:
Yep, otherwise the tests fail.

Reply:
> Yep, otherwise the tests fail.

… or the test is buggy. 😅 I think the question is what `TxMetaBytes` means: Is it the hex-encoding of a sequence of bytes, or is it the raw sequence of bytes? I would bet that the `Cardano.Api` understands it as a raw sequence of bytes.

I suppose that we can only find that out by comparing at least one snippet of

* metadata CBOR received through the cardano-node
* metadata CBOR received through Blockfrost

against `Cardano.Api.deserialiseFromCBOR`, I suppose. 🤔

Reply:
Yes